### PR TITLE
[IR] Add disjoint flag for Or instructions.

### PR DESCRIPTION
**Pull request by topperc**
This flag indicates that every bit is known to be zero in at least one of the inputs. This allows the Or to be treated as an Add since there is no possibility of a carry from any bit.

If the flag is present and this property does not hold, the result is poison.

This makes it easier to reverse the InstCombine transform that turns Add into Or.

This is inspired by a comment here https://github.com/llvm/llvm-project/pull/71955#discussion_r1391614578

Discourse thread https://discourse.llvm.org/t/rfc-add-or-disjoint-flag/75036

### CODE REVIEW

File: llvm/include/llvm/IR/InstrTypes.h
Comment (nikic) [inline llvm/include/llvm/IR/InstrTypes.h:1104]: Dropped character

**Review by nikic — COMMENTED** [review]
This basically looks fine to me.

This is missing the flag drop in InstCombineSimplifyDemanded.

Can you please also add a test to llvm/test/Transforms/SimplifyCFG/HoistCode.ll to cover the flag intersect?

File: llvm/docs/LangRef.rst
Comment (nhaehnle) [inline llvm/docs/LangRef.rst:10020]: I stumbled over this a little. Perhaps it should be "... is a poison value if both inputs have a one in the same position"? (I'm not 100% happy about that either)

File: llvm/include/llvm/IR/Instruction.h
Comment (nhaehnle) [inline llvm/include/llvm/IR/Instruction.h:454]: Mirroring a remark on https://github.com/llvm/llvm-project/pull/72501: Perhaps this method should only exist on `PossiblyDisjointInst`? (and same for `isDisjoint` below)

**Review by nhaehnle — COMMENTED** [review]
Nice! I'm in favor (as should be obvious from my earlier comments...)

File: llvm/include/llvm/IR/InstrTypes.h
Comment (nikic) [inline llvm/include/llvm/IR/InstrTypes.h:418]:
```suggestion
/// An or instruction, which can be marked as "disjoint", indicating that the
```
(I think?)

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:236]: Can you please also add a test for this?

File: llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp:258]: As long as this sticks with shrinking only, we shouldn't have to drop flags here. (It can only get "more" disjoint.)

File: llvm/include/llvm/IR/InstrTypes.h
Comment (topperc) [reply llvm/include/llvm/IR/InstrTypes.h:418]: Yes

File: llvm/test/Transforms/InstCombine/or.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/or.ll:1581]: Missing CHECK lines?

**Review by nikic — APPROVED** [review]
LGTM, but please wait for an additional approval.

File: llvm/test/Bitcode/flags.ll
Comment (arsenm) [inline llvm/test/Bitcode/flags.ll:1]: Should there also be a test in test/Bitcode/compatibility?

File: llvm/docs/LangRef.rst
Comment (arsenm) [inline llvm/docs/LangRef.rst:10016]: My reading of "every bit" seems a bit ambiguous. I thought the point was that comparing each individual bit, each must be 0. But this reads to me more like one of the values as a whole is 0

Comment (topperc) [reply llvm/docs/LangRef.rst:10016]: That was not my intention. I'll try to reword it.

**Review by nhaehnle — COMMENTED** [review]
This mostly LGTM, but I see two questions that haven't been addressed:

* Should a test be added to Bitcode/compatibility (this seems not to have been done consistently in recent times, but the answer is probably still yes)
* Should the setter/getter be moved to PossiblyDisjointInst (the other discussion does suggest that this should probably be done)

**Comment by nikic** [comment]
> This mostly LGTM, but I see two questions that haven't been addressed:
> 
>     * Should a test be added to Bitcode/compatibility (this seems not to have been done consistently in recent times, but the answer is probably still yes)

I'd probably say no on this one, because the "release process" for that file has not been followed since LLVM 6, for more than ten releases. We should probably just remove the relevant section of the docs.

> * Should the setter/getter be moved to PossiblyDisjointInst (the other discussion does suggest that this should probably be done)

I'm neutral on this one. The current implementation matches existing practice, but I wouldn't mind giving the more principled design a try either.

File: llvm/test/Bitcode/compatibility.ll
Comment (nikic) [inline llvm/test/Bitcode/compatibility.ll:1362]: Missing CHECK line.

File: llvm/lib/IR/Instruction.cpp
Comment (nikic) [inline llvm/lib/IR/Instruction.cpp:321]: I believe the suggestion was to not have these methods on Instruction, only the ones on PossiblyDisjointInst.

**Comment by nhaehnle** [comment]
> > This mostly LGTM, but I see two questions that haven't been addressed:
> > 
> > * Should a test be added to Bitcode/compatibility (this seems not to have been done consistently in recent times, but the answer is probably still yes)
> 
> I'd probably say no on this one, because the "release process" for that file has not been followed since LLVM 6, for more than ten releases. We should probably just remove the relevant section of the docs.

Fair enough.

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:10016]:
```suggestion
``disjoint`` means that for each bit, that bit is zero in at least one of the inputs.
```
I think?

**Review by nikic — APPROVED** [review]

**Review by nhaehnle — APPROVED** [review]
Thanks!

**Comment by nikic** [comment]
Fix for an InstSimplify miscompile with `disjoint`: https://github.com/llvm/llvm-project/commit/07c18a05e2cb80e87b0c681c147a253565064f3e

**Comment by nikic** [comment]
Another miscompilation fix: https://github.com/llvm/llvm-project/commit/ea602cb80678fbf5f52bbbda142437164b2a3407 Should really have caught this one during review :/

**Comment by preames** [comment]
I'm seeing a downstream assertion failure building SPEC which I suspect is triggered by this change.

```
LoopVectorize.cpp:1122: llvm::InnerLoopVectorizer::collectPoisonGeneratingRecip
es(llvm::VPTransformState&)::<lambda(llvm::VPRecipeBase*)>: Assertion `(!Instr || !Instr->hasPoisonGeneratingFlags()) && "found inst
ruction with poison generating flags not covered by " "VPRecipeWithIRFlags"' failed.
```
I can reduce a test case if needed, but the assertion failure seems fairly self explanatory.

**Comment by alexey-bataev** [comment]
> I'm seeing a downstream assertion failure building SPEC which I suspect is triggered by this change.
> 
> ```
> LoopVectorize.cpp:1122: llvm::InnerLoopVectorizer::collectPoisonGeneratingRecip
> es(llvm::VPTransformState&)::<lambda(llvm::VPRecipeBase*)>: Assertion `(!Instr || !Instr->hasPoisonGeneratingFlags()) && "found inst
> ruction with poison generating flags not covered by " "VPRecipeWithIRFlags"' failed.
> ```
> 
> I can reduce a test case if needed, but the assertion failure seems fairly self explanatory.

Most probably it should be fixed by this https://github.com/llvm/llvm-project/pull/74112

**Comment by fhahn** [comment]
> I can reduce a test case if needed, but the assertion failure seems fairly self explanatory.

It might be a separate issue to #74112, so if it would be possible to reduce a test case to check if it is indeed an `or disjoint` that's causing this, that would be helpful. In any case, we should support preserving the new `disjoint` flag, so I put up #74364

**Comment by nikic** [comment]
Do you plan to propagate this down to SDAG?

**Comment by topperc** [comment]
> Do you plan to propagate this down to SDAG?

Yes. I have a prototype patch. I'll post it soon.

**Pull request merged**
